### DISCUSSION TITLE
How to correctly set the middleware for authentication.

### DISCUSSION THREAD

**Discussion by emflomed17** [vercel/next.js#62364] [Help]
### Summary

I'm really struggling trying to set my auth middleware. 

In the oficial docs says the following: 

> Not all routes require authorization. Use the matcher option in your Middleware to specify any routes that do not require authorization checks.

But it seems is doing the opposite.

The middleware is being called on routes defined in matcher.

middleware.ts
```
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";
import { jwtVerify } from "jose";

export async function middleware(request: NextRequest) {
  const jwtCookie = request.cookies.get("cookie_token");

  if (!jwtCookie)
    return NextResponse.redirect(new URL("/auth/login", request.url));

  const jwt = jwtCookie.value; // Extracting the JWT token string

  // this condition avoid to show the login page if the user is logged in
  if (jwt) {
    if (request.nextUrl.pathname.includes("/login")) {
      try {
        await jwtVerify(jwt, new TextEncoder().encode("this_is_my_secret"));
        return NextResponse.redirect(new URL("/chat", request.url));
      } catch (error) {
        return NextResponse.next();
      }
    }
  }

  try {
    const { payload } = await jwtVerify(
      jwt,
      new TextEncoder().encode("this_is_my_secret")
    );

    return NextResponse.next();
  } catch (error) {
    console.log(error);

    return NextResponse.redirect(new URL("/auth/login", request.url));
  }
}

export const config = {
  matcher: ["/chat/:path*"],
};

```
Could say that the middleware in certain way is protecting the routes correctly. If the cookie is not valid or is not present and I try to navigate to routes under `/chat` I'm redirected to `/auth/login`

The problem is I can not redirect to `/chat` routes when the cookie is valid because this middleware is never called in `/auth/login` route

The cookie is being set correctly but I'm not redirected. 

Please help me to correctly set my matcher.

DETAILS: 
 "next": "^14.1.0"
 
 folder structure:
 ![Screenshot 2024-02-21 at 7 10 19 p m](https://github.com/vercel/next.js/assets/25670760/e131b4de-98c2-4fe9-b786-fa091b00ecb3)

**Comment by emflomed17** [comment]
Anyone? 👀

**Comment by ananuness** [comment]
Hi!! Let me see if I understood it... You want to protect all your routes under `/chat/*`, but when the user already has a cookie set and you try to access the `/auth/login` you want to redirect him to the `/chat`, right? 

If so, you have to put your middleware in all your routes (removing the matcher), because the way it's set in your `matcher` now, it's only called when the user try to access `/chat/:path*`. As you're verifying if the user is logged in your middleware, there's no problem to let it cover all your routes.
